src/Network/Tox/Application/GroupChatsPackets.lhs

```diff
@@ -0,0 +1,654 @@
+  Length                & Contents \\
+  \hline
+  \texttt{2}            & Peerlist Checksum \\
+  \texttt{2}            & Confirmed Peer Count \\
+  \texttt{4}            & Shared State Version \\
+  \texttt{4}            & Sanctions Credentials Version \\
+  \texttt{2}            & Sanctions Credentials Checksum \\
+  \texttt{4}            & Topic Version \\
+  \texttt{2}            & Topic Checksum \\
+  Variable              & Packed IP Address and Port \\
+\end{tabular}
+
+Ping packets are periodically sent to every confirmed peer in order to
+maintain peer connections, and to ensure the group state between peers
+are in sync. A peer is considered to be disconnected from the group after
+a ping packet has not been receieved over a period of time.
+
+\subsection{MESSAGE\_ACK (0x02)}
+
+Message ack packets are structured as follows:
+
+\begin{tabular}{l|l}
+  Length                & Contents \\
+  \hline
+  \texttt{8}            & Message ID \\
+  \texttt{1}            & Type \\
+\end{tabular}
+
+This packet ensures that all lossless packets are successfully
+received and processed in sequential order as they were sent.
+
+Message ack types are defined by an enumerator beginning at zero as
+follows:
+
+\begin{tabular}{l|l}
+  Type                          & ID \\
+  \hline
+  \textbf{\verb'GR_ACK_RECV'}   & 0x00 \\
+  \textbf{\verb'GR_ACK_REQ'}    & 0x01 \\
+\end{tabular}
+
+If the type is \textbf{\verb'GR_ACK_RECV'}, this indicates that the packet
+with the given id has been received and successfully processed. If the
+type is \textbf{\verb'GR_ACK_REQ'}, this indicates that the message with the
+given id should be sent again.
+
+\subsection{INVITE\_RESPONSE\_REJECT (0x03)}
+
+An invite response reject payload is structured as follows:
+
+\begin{tabular}{l|l}
+  Length                & Contents \\
+  \hline
+  \texttt{1}            & Type \\
+\end{tabular}
+
+This packet alerts a peer that their invite request has been rejected. The
+reason for the rejection is specified by the \textbf{\verb'type'} field.
+
+Rejection types are defined by an enumerator beginning at zero as
+follows:
+\begin{tabular}{l|l}
+  Type                              & ID \\
+  \hline
+  \textbf{\verb'GROUP_FULL'}        & 0x00 \\
+  \textbf{\verb'INVALID_PASSWORD'}  & 0x01 \\
+  \textbf{\verb'INVITE_FAILED'}     & 0x02 \\
+\end{tabular}
+
+\section{Lossless Packet Payloads}
+
+\subsection{KEY\_ROTATIONS (0xf0)}
+
+Key rotation packets are structured as follows:
+
+\begin{tabular}{l|l}
+  Length                & Contents \\
+  \hline
+  \texttt{1}            & \textbf{\verb'is_response'} \\
+  \texttt{32}           & Public Encryption Key \\
+\end{tabular}
+
+Key rotation packets are used to rotate session encryption keys with
+a peer. If \textbf{\verb'is_response'} is false, the packet initiates a
+public key exchange. Otherwise the packet is a response to a previously
+initiated exchange.
+
+The public encryption key must be a brand new, unused key, which takes
```

Comment:
"brand new, unused" could also be "new" or "newly generated", no? "brand new" sounds dramatic :P